subsys/zbus/zbus.c

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (c) 2022 Rodrigo Peixoto <rodrigopex@gmail.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/zbus/zbus.h>
+LOG_MODULE_REGISTER(zbus, CONFIG_ZBUS_LOG_LEVEL);
+
+#if defined(CONFIG_THREAD_NAME)
+#define _ZBUS_CURRENT_THREAD_INFO k_thread_name_get(k_current_get()), k_current_get()
+#else
+#define _ZBUS_CURRENT_THREAD_INFO "_", k_current_get()
+#endif /* CONFIG_THREAD_NAME */
+
+inline k_timeout_t _zbus_timeout_remainder(uint64_t end_ticks)
+{
+	int64_t now_ticks = sys_clock_tick_get();
+
+	return K_TICKS((k_ticks_t)MAX(end_ticks - now_ticks, 0));
+}
+
+void __weak _zbus_notify_runtime_listeners(struct zbus_channel *chan)
+{
+}
+
+int __weak _zbus_notify_runtime_subscribers(struct zbus_channel *chan, uint64_t end_ticks)
+{
+	return 0;
+}
+
+void _zbus_notify_listeners(struct zbus_channel *chan)
+{
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled && ((*obs)->callback != NULL)) {
+			(*obs)->callback(chan);
```

Comment:
Have you considered using separate arrays of listeners and subscribers? Currently, the observers array is always iterated twice on each publishing to the channel, even when there is no listeners or no subscribers.

Reply:
For runtime observers, it will be necessary to separate as well to keep consistency. Do you think it would be interesting? We will need to have  `add_sub` and `add_lis` instead of `add_obs`. @anangl